trin-core/src/lib.rs

```diff
@@ -6,3 +6,137 @@ pub mod jsonrpc;
 pub mod portalnet;
 pub mod socket;
 pub mod utils;
+
+pub const ACQUIRE_TIMEOUT_MS: u64 = 100;
+pub const HOLD_TIMEOUT_MS: u64 = 100;
+
+pub struct TimedRwReadGuard<'a, T: ?Sized> {
+    pub inner: tokio::sync::RwLockReadGuard<'a, T>,
+    pub acquisition_line: u32,
+    pub acquisition_file: &'static str,
+    pub acquisition_time: std::time::Instant,
+    pub sleep_task: tokio::task::JoinHandle<()>,
+}
+
+async fn sleep_then_log(file: &'static str, line: u32) {
+    tokio::time::sleep(std::time::Duration::from_millis(HOLD_TIMEOUT_MS)).await;
+    log::warn!(
+        "[{}:{}] lock held for over {}ms, not yet released",
+        file,
+        line,
+        HOLD_TIMEOUT_MS.to_string()
+    );
+}
+
+impl<'a, T: ?Sized> TimedRwReadGuard<'a, T> {
+    pub async fn new(
+        inner: tokio::sync::RwLockReadGuard<'a, T>,
+        acquisition_line: u32,
+        acquisition_file: &'static str,
+    ) -> TimedRwReadGuard<'a, T> {
+        let now = std::time::Instant::now();
+        let move_line = acquisition_line;
+        let move_file = acquisition_file;
+        let handle = tokio::spawn(async move {
+            sleep_then_log(move_file, move_line).await;
+        });
+
+        TimedRwReadGuard {
+            inner,
+            acquisition_line,
+            acquisition_file,
+            acquisition_time: now,
+            sleep_task: handle,
+        }
+    }
+}
+
+impl<'a, T> std::ops::Deref for TimedRwReadGuard<'a, T> {
+    type Target = tokio::sync::RwLockReadGuard<'a, T>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<'a, T: ?Sized> Drop for TimedRwReadGuard<'a, T> {
+    fn drop(&mut self) {
+        self.sleep_task.abort();
+        let held_for = self.acquisition_time.elapsed().as_millis();
+        if held_for > HOLD_TIMEOUT_MS.into() {
+            log::warn!(
+                "[{}:{}] lock held for too long: {}ms",
+                self.acquisition_file,
+                self.acquisition_line,
+                held_for,
+            )
+        }
+    }
+}
+
+#[macro_export]
+macro_rules! rw_read {
```

Comment:
If we're going to leave this lock macro in place everywhere (which I'm cool with), then we should probably only run this tooling overhead on non-release builds. Something like branching on `debug_assertions` (I found that trick in [this post](https://stackoverflow.com/a/39205417/8412986)).

Reply:
I experimented with using `debug_assertions`: https://github.com/ethereum/trin/pull/141/commits/0cb24458a48077882f49bc69352a2494516b3c8a

I think I'd prefer to drop this commit though. It makes the code a fair amount less readable, doesn't seem like it drops very much overhead, and prevents us from being able to diagnose deadlocks if someone running a release build tries to send us their logs.

What do you think?